packages/core/lib/patchers/aws3_p.ts

```diff
@@ -0,0 +1,173 @@
+import type {
+  MetadataBearer,
+  ResponseMetadata,
+  Command,
+  Client,
+} from '@aws-sdk/types';
+
+import type { RegionInputConfig } from '@aws-sdk/config-resolver';
+
+import { isThrottlingError } from '@aws-sdk/service-error-classification';
+
+import ServiceSegment from '../segments/attributes/aws';
+
+import { stringify } from 'querystring';
+
+import Subsegment from '../segments/attributes/subsegment';
+
+var contextUtils = require('../context_utils');
+
+var logger = require('../logger');
+
+import { getCauseTypeFromHttpStatus } from '../utils';
+import { SdkError } from '@aws-sdk/smithy-client';
+import { SegmentLike } from '../aws-xray';
+
+type HttpResponse = { response: { status: number } };
+
+async function buildAttributesFromMetadata(
+  client: any,
+  command: any,
+  metadata: ResponseMetadata,
+): Promise<[ServiceSegment, HttpResponse]> {
+  const { extendedRequestId, requestId, httpStatusCode: statusCode, attempts } = metadata;
+  const serviceIdentifier = client.config.signingName as string;
+
+  let operation: string = command.constructor.name.slice(0, -7);
+
+  const aws = new ServiceSegment(
+    {
+      extendedRequestId,
+      requestId,
+      retryCount: attempts,
+      request: {
+        operation,
+        httpRequest: {
+          region: await client.config.region(),
+          statusCode,
+        },
+        params: command.input,
+      },
+    },
+    serviceIdentifier,
+  );
+
+  const http = { response: { status: statusCode || 0 } };
+  return [aws, http];
+}
+
+function addFlags(http: HttpResponse, subsegment: Subsegment, err?: SdkError): void {
+  if (err && isThrottlingError(err)) {
+    subsegment.addThrottleFlag();
+  } else if (http.response?.status === 429) {
+    subsegment.addThrottleFlag();
+  }
+
+  const cause = getCauseTypeFromHttpStatus(http.response.status);
+  if (cause === 'fault') {
+    subsegment.addFaultFlag();
+  } else if (cause === 'error') {
+    subsegment.addErrorFlag();
+  }
+}
+
+type DefaultConfiguration = RegionInputConfig & {
+  signingName: string;
+};
+
+export function captureAWSClient<
+  Input extends object,
+  Output extends MetadataBearer,
+  Configuration extends DefaultConfiguration
+>(client: Client<Input, Output, Configuration>, manualSegment?: SegmentLike): Client<Input, Output, Configuration> {
+  // create local copy so that we can later call it
+  const send = client.send;
+
+  const serviceIdentifier = client.config.signingName;
```

Comment:
fyi I opened a feature request to the aws-sdk-js-v3 for an easier way to get service/operation names: https://github.com/aws/aws-sdk-js-v3/issues/2030

Reply:
we will need to require at least version 3.2.0 of the AWS V3 SDK. Otherwise we'll need to detect the version of the SDK and run dedicated code to detect the version. Also, the metadata object has changes with 3.2.0 so the changes would be quite significant to also support 3.1.0.